packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +540,693 @@ class ShellRoute extends RouteBase {
+
+  /// The observers for this branch.
+  ///
+  /// The observers parameter is used by the [Navigator] built for this branch.
+  final List<NavigatorObserver>? observers;
+}
+
+/// Builder for a custom container for shell route Navigators.
+typedef ShellNavigatorContainerBuilder = Widget Function(BuildContext context,
+    StackedShellRouteState shellRouteState, List<Widget> children);
+
+/// Widget for managing the state of a [StackedShellRoute].
+///
+/// Normally, this widget is not used directly, but is instead created
+/// internally by StackedShellRoute. However, if a custom container for the
+/// branch Navigators is required, StackedNavigationShell can be used in
+/// the builder or pageBuilder methods of StackedShellRoute to facilitate this.
+/// The container is created using the provided [ShellNavigatorContainerBuilder],
+/// where the List of Widgets represent the Navigators for each branch.
+///
+/// Example:
+/// ```
+/// builder: (BuildContext context, StackedShellRouteState state, Widget child) {
+///   return StackedNavigationShell(
+///     shellRouteState: state,
+///     containerBuilder: (_, __, List<Widget> children) => MyCustomShell(shellState: state, children: children),
+///   );
+/// }
+/// ```
+class StackedNavigationShell extends StatefulWidget {
+  /// Constructs an [_StackedNavigationShell].
+  StackedNavigationShell({
+    required this.shellRouteState,
+    ShellNavigatorContainerBuilder? containerBuilder,
+  })  : containerBuilder = containerBuilder ?? _defaultChildBuilder,
+        super(key: shellRouteState._shellStateKey);
+
+  static Widget _defaultChildBuilder(BuildContext context,
+      StackedShellRouteState shellRouteState, List<Widget> children) {
+    return _IndexedStackedRouteBranchContainer(
+        currentIndex: shellRouteState.currentIndex, children: children);
+  }
+
+  /// The current state of the associated [StackedShellRoute].
+  final StackedShellRouteState shellRouteState;
+
+  /// The builder for a custom container for shell route Navigators.
+  final ShellNavigatorContainerBuilder containerBuilder;
+
+  @override
+  State<StatefulWidget> createState() => StackedNavigationShellState();
+}
+
+/// State for StackedNavigationShell.
+class StackedNavigationShellState extends State<StackedNavigationShell>
+    with RestorationMixin {
+  final Map<Key, Widget> _branchNavigators = <Key, Widget>{};
+
+  StackedShellRoute get _route => widget.shellRouteState.route;
+
+  StackedShellRouteState get _routeState => widget.shellRouteState;
+
+  GoRouter get _router => _routeState._router;
+  RouteMatcher get _matcher => _router.routeInformationParser.matcher;
+  GoRouteInformationProvider get _routeInformationProvider =>
+      _router.routeInformationProvider;
+
+  final Map<StackedShellBranch, _RestorableRouteMatchList> _branchLocations =
+      <StackedShellBranch, _RestorableRouteMatchList>{};
+
+  @override
+  String? get restorationId => _route.restorationScopeId;
```

Comment:
This should probably be pass into StackedNavigationShell as constructor parameter.